factory-containers/compose_app_downloader.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/python3
+
+import yaml
+import subprocess
+import logging
+import time
+import os
+import tempfile
+import shutil
+import argparse
+import contextlib
+
+
+logger = logging.getLogger(__name__)
+
+
+class ComposeApp:
+    def __init__(self, name, file_path):
+        self.name = name
+        self.file_path = file_path
+
+        with open(self.file_path) as compose_file:
+            self._composed_app = yaml.safe_load(compose_file)
+
+    def services(self):
+        return self._composed_app['services'].items()
+
+
+class ContainerDaemon:
+    def __init__(self):
+        self.cmd = 'containerd'
+
+    def __enter__(self):
+        self._container_dir = tempfile.mkdtemp()
+
+        self._root_dir = os.path.join(self._container_dir, 'root')
+        os.mkdir(self._root_dir)
+
+        self._state_dir = os.path.join(self._container_dir, 'state')
+        os.mkdir(self._state_dir)
+
+        self.address = os.path.join(self._container_dir, 'containerd.sock')
+
+        logger.info("Starting containerd for docker daemon...")
+        self._process = subprocess.Popen([self.cmd, '--address', self.address,
+                                          '--root', self._root_dir,
+                                          '--state', self._state_dir],
+                                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # hack, let containerd some time to start
+        time.sleep(2)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._process.terminate()
+        self._process.wait(timeout=60)
+
+        logger.info("Containerd daemon has been stopped")
+        shutil.rmtree(self._container_dir)
+
+
+class DockerDaemon:
+    ImageDataDirs = ['overlay2', 'image']
+
+    def __init__(self, dst_dir: str, containerd_addr: str):
+        self.data_root = dst_dir
+        self._containerd_addr = containerd_addr
+        #self.cmd = '/home/mike/bin/dockerd-rootless.sh'
```

Comment:
Please remove this line if not really required.

Reply:
> Just some minor comments, but LGTM, will let @doanac do the final +1 here.

Updated (forced push) according to the comments.